crates/common/mqtt_channel/src/session.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{Config, Connection, MqttError};
+use rumqttc::{AsyncClient, Event, Packet};
+
+/// Create a persistent session on the MQTT server `config.host`.
+///
+/// The session is named after the `config.session_name`
+/// subscribing to all the topics given by the `config.subscriptions`.
+///
+/// A new `Connection` created with a config with the same session name,
+/// will receive all the messages published meantime on the subscribed topics.
+///
+/// This function can be called multiple times with the same session name,
+/// since it consumes no messages.
+pub async fn init_session(config: &Config) -> Result<(), MqttError> {
+    if config.clean_session || config.session_name.is_none() {
+        return Err(MqttError::InvalidSessionConfig);
+    }
+
+    let mqtt_options = config.mqtt_options();
+    let (mqtt_client, mut event_loop) = AsyncClient::new(mqtt_options, config.queue_capacity);
+
+    loop {
+        match event_loop.poll().await {
+            Ok(Event::Incoming(Packet::ConnAck(_))) => {
+                let subscriptions = config.subscriptions.filters();
+                if subscriptions.is_empty() {
+                    break;
+                }
+                mqtt_client.subscribe_many(subscriptions).await?;
+            }
+
+            Ok(Event::Incoming(Packet::SubAck(_))) => {
+                break;
+            }
+
+            Err(err) => {
+                if Connection::pause_on_error(&err) {
+                    Connection::do_pause().await;
+                }
+            }
+            _ => (),
+        }
+    }
+
+    let _ = mqtt_client.disconnect().await;
```

Comment:
What happens if an `Err` is returned here?

What can a caller of the function be assured of that could have happened?


If I call `init_session` and then rely on that session to be active, will that still be true if the disconnection fails?

Reply:
I think this error can be ignored safely and that it would on the contrary misleading to report it.

* Disconnect sends a [`DISCONNECT`](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718090). No response is expected from the server.
* The only reason this can fail is that the connection has already been lost.
* This has no impact on the state of the session that has already been created and will stay active.

That said, error checking can be improved, not on disconnect, but on reception of the [`CONNACK`](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718033). We have to check that the connection has been actually accepted. A check that is currently missing.

=> I will improve error handling on `CONNACK`.